### PR TITLE
s3_bucket: add encryption capabilities to the module 

### PR DESCRIPTION
**Pull request by stefanhorning**
##### SUMMARY
Adds encryption capabilities to the s3_bucket module. Based on this PR #38538

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
s3_bucket

##### ADDITIONAL INFORMATION
Includes all change sfrom @julienvey made in PR #38538 and the changes from @wimnat made in his branch https://github.com/ansible/ansible/compare/devel...wimnat:s3_encryption?expand=1 and resolving of merge conflicts by me. To hopefully get this all into a mergable state.

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/s3_bucket.py [hunk 1]
```diff
@@ -555,11 +631,13 @@ def main():
...
             versioning=dict(default=None, type='bool'),
-            ceph=dict(default='no', type='bool')
+            ceph=dict(default='no', type='bool'),
+            encryption=dict(choices=['none', 'AES256', 'aws:kms']),
+            encryption_key_id=dict()
         )
     )
 
-    module = AnsibleAWSModule(argument_spec=argument_spec)
+    module = AnsibleAWSModule(argument_spec=argument_spec, required_if=[['encryption', 'aws:kms', ['encryption_key_id']]])
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/s3_bucket.py]: better to split this onto two lines

**Comment by willthames** [comment]
Looks good, I'll run the test suite

**Comment by stefanhorning** [comment]
I rebased this branch once more and resolved the last merging glitches now. @willthames also added the requested linebreak. So I guess this is ready to merge now.

**Comment by stefanhorning** [comment]
I think s.b. needs to restart the build here (I didn't find a button on the shippable UI) as the failing test was working in the previous commit before I fixed the last remaining test for the linting, so technically the build is green here.

**Comment by stefanhorning** [comment]
cc @wimnat @ryansb @s-hertel (previous reviewers)

**Comment by stefanhorning** [comment]
ready_for_review

File: lib/ansible/modules/cloud/amazon/s3_bucket.py [hunk 2]
```diff
@@ -555,11 +636,15 @@ def main():
...
-            ceph=dict(default='no', type='bool')
+            ceph=dict(default='no', type='bool'),
+            encryption=dict(choices=['none', 'AES256', 'aws:kms']),
+            encryption_key_id=dict()
         )
     )
 
-    module = AnsibleAWSModule(argument_spec=argument_spec)
+    module = AnsibleAWSModule(
+        argument_spec=argument_spec, required_if=[['encryption', 'aws:kms', ['encryption_key_id']]]
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/s3_bucket.py]:
```suggestion
        argument_spec=argument_spec, 
        required_if=[['encryption', 'aws:kms', ['encryption_key_id']]]
```
The line break should separate the args, see the style of other aws modules for examples.

**Review by jillr — CHANGES_REQUESTED** [review]
tests good for me, thanks for this patch.

**Comment by stefanhorning** [comment]
@jillr Added the linebreak as suggested.

To everyone else: if there is no more linebreaks to add, please merge.

**Review by jillr — APPROVED** [review]

**Review by jillr — COMMENTED** [review]
Once sanity tests pass, lgtm

**Comment by stefanhorning** [comment]
Sanity tests are green now. But somehow integration tests started failing out of the blue. Was there recent change in the test environment in regards to AWS s3 permissions? If so they need to be rolled back to make the tests in here green again.

**Comment by stefanhorning** [comment]
Rebased on latest devel to see if that fixes build issues.

**Comment by stefanhorning** [comment]
Ready to merge now @wimnat @ryansb @s-hertel @jillr @willthames

**Pull request merged**